app/plugin_hooks_test.go

```diff
@@ -662,6 +662,49 @@ func TestHookFileWillBeUploaded(t *testing.T) {
 	})
 }
 
+func TestHookNotificationWillBeSent(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	tearDown, pluginIds, _ := SetAppEnvironmentWithPlugins(t,
+		[]string{
+			`
+		package main
+
+		import (
+			"github.com/mattermost/mattermost-server/model"
+			"github.com/mattermost/mattermost-server/plugin"
+		)
+
+		type MyPlugin struct {
+			plugin.MattermostPlugin
+		}
+
+		func (p *MyPlugin) NotificationWillBeSent(c *plugin.Context, post *model.Post, userList map[string]bool) map[string]bool {
+			delete(userList, "user1")
+			userList["user3"] = false
```

Comment:
Please also test adding a user

Reply:
`userList["user3"] = false` actually adds a new user.  
The value is only used by [sendPushNotification](https://github.com/mattermost/mattermost-server/blob/484400ef034d9711a5eb33c37015e1fa7917abbd/app/notification.go#L273) and should be true if it's an explicit mention, false otherwise.

Reply:
Oh, I see. Thanks for pointing this out :+1: 

Would you please improve the documentation of `NotificationWillBeSent` to make this distinction clear?

Reply:
I just noticed there have been some big changes in master's `app/notification.go`.  
The mentionedUserIds list is now a `&ExplicitMentions{}` taking the userId and a mentionType.  
Most of my code has to be rewritten, I'll look at this shortly.

Reply:
Sorry for this :/